include/proxy-wasm/context.h

```diff
@@ -0,0 +1,736 @@
+   * serialized..
+   */
+  virtual WasmResult setProperty(string_view /* key */, string_view /* value */) {
+    unimplemented();
+    return WasmResult::Unimplemented;
+  }
+
+  // Continue
+
+  /**
+   * Continue processing a request e.g. after returning ProxyAction::Pause.
+   */
+  virtual void continueRequest() { unimplemented(); }
+
+  /**
+   * Continue processing a response e.g. after returning ProxyAction::Pause.
+   */
+  virtual void continueResponse() { unimplemented(); }
+
+  /**
+   * Respond directly to an HTTP request.
+   * @param response_code is the response code to send.
+   * @param body is the body of the response.
+   * @param additional_headers are additional headers to send in the response.
+   * @param grpc_status is an optional gRPC status if the connection is a gRPC connection.
+   * @param details are details of any (gRPC) error.
+   */
+  virtual void sendLocalResponse(uint64_t /* response_code */, string_view /* body */,
+                                 Pairs /* additional_headers */, uint64_t /* grpc_status */,
+                                 string_view /* details */) {
+    unimplemented();
+  }
+
+  // Shared Data
+
+  /**
+   * Get proxy-wide key-value data shared between VMs.
+   * @param key is a proxy-wide key mapping to the shared data value.
+   * @param cas is a number which will be incremented when a data value has been changed.
+   * @param data is a location to store the returned value.
+   */
+  virtual WasmResult getSharedData(string_view /* key */,
+                                   std::pair<std::string, uint32_t /* cas */> * /* data */);
+
+  /**
+   * Set a key-value data shared between VMs.
+   * @param key is a proxy-wide key mapping to the shared data value.
+   * @param cas is a compare-and-swap value. If it is zero it is ignored, otherwise it must match
+   * the cas associated with the value.
+   * @param data is a location to store the returned value.
+   */
+  virtual WasmResult setSharedData(string_view /* key */, string_view /* value */,
+                                   uint32_t /* cas */);
+
+  // Shared Queue
+
+  /**
+   * Register a proxy-wide queue.
+   * @param queue_name is a name for the queue. The queue_name is combined with the vm_id (if any)
+   * to make a unique identifier for the queue.
+   * @param token_ptr a location to store a token corresponding to the queue.
+   */
+  virtual WasmResult registerSharedQueue(string_view /* queue_name */, uint32_t *token_ptr);
+
+  /**
+   * Get the token for a queue.
+   * @param vm_id is the vm_id of the Plugin of the Root Context which registered the queue.
+   * @param queue_name is a name for the queue. The queue_name is combined with the vm_id (if any)
+   * to make a unique identifier for the queue.
+   * @param token_ptr a location to store a token corresponding to the queue.
+   */
+  virtual WasmResult resolveSharedQueue(string_view /* vm_id */, string_view /* queue_name */,
```

Comment:
When does someone use resolve() vs. register()? Couldn't this be combined into a single function with a parameter that says whether to "create if not exist" or similar?

Reply:
Register is used to create a queue for reading from.
Resolve is used to get a reference to a previously created queue for the purposes of writing to.

Reply:
These are called by different VMs.

Reply:
Even if they are called on different VMs I'm not sure why we need 2 different functions here that basically do the same thing? If you want to keep them separate I might consider switching "resolve" to "lookup" to be more clear, but up to you.

Reply:
The do not do the same thing.  One creates a queue for reading.  The other gets a reference to an existing queue for writing.   That is like saying that accept() and connect() do the same thing.   Yes they both return a socket, but that is where the similarity ends.

Reply:
Let's take this so the spec then. If they do different things the naming should be more clear such as `registerForRead()` or `registerForWrite`. Even in that case, they can still be a single function with a param that specifies the read/write direction.

Reply:
+1 on preferring `lookup`.

Reply:
Sure, lookup is reasonable.   I am kind of surprised by the desire to combine these while splitting apart the metric calls.  To me these are completely different whereas the metrics calls are very similar.